<a href="https://colab.research.google.com/github/19PA1A0470/AI-LAB/blob/master/adaboost.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **ADA boost model**

In [0]:
# importimg required libraries
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix


In [0]:
# reading data
ipl=pd.read_csv("Iris.csv")
ipl

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

In [0]:
ipl.isna().any() # to find NA's

Id               False
SepalLengthCm    False
SepalWidthCm     False
PetalLengthCm    False
PetalWidthCm     False
Species          False
dtype: bool

In [0]:
#converting catorical feature to numerical feature
ipl['Species'].replace(to_replace=['Iris-setosa','Iris-versicolor','Iris-virginica'], value=[0,1,2],inplace=True)
ipl.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  Species
0   1            5.1           3.5            1.4           0.2        0
1   2            4.9           3.0            1.4           0.2        0
2   3            4.7           3.2            1.3           0.2        0
3   4            4.6           3.1            1.5           0.2        0
4   5            5.0           3.6            1.4           0.2        0

In [0]:
# declaring featuresa and target variable
X = ipl[['SepalWidthCm','PetalLengthCm','PetalWidthCm','SepalLengthCm']].values
y= ipl['Species'].values
print('X=',X)
print('y=',y)

X= [[3.5 1.4 0.2 5.1]
 [3.  1.4 0.2 4.9]
 [3.2 1.3 0.2 4.7]
 [3.1 1.5 0.2 4.6]
 [3.6 1.4 0.2 5. ]
 [3.9 1.7 0.4 5.4]
 [3.4 1.4 0.3 4.6]
 [3.4 1.5 0.2 5. ]
 [2.9 1.4 0.2 4.4]
 [3.1 1.5 0.1 4.9]
 [3.7 1.5 0.2 5.4]
 [3.4 1.6 0.2 4.8]
 [3.  1.4 0.1 4.8]
 [3.  1.1 0.1 4.3]
 [4.  1.2 0.2 5.8]
 [4.4 1.5 0.4 5.7]
 [3.9 1.3 0.4 5.4]
 [3.5 1.4 0.3 5.1]
 [3.8 1.7 0.3 5.7]
 [3.8 1.5 0.3 5.1]
 [3.4 1.7 0.2 5.4]
 [3.7 1.5 0.4 5.1]
 [3.6 1.  0.2 4.6]
 [3.3 1.7 0.5 5.1]
 [3.4 1.9 0.2 4.8]
 [3.  1.6 0.2 5. ]
 [3.4 1.6 0.4 5. ]
 [3.5 1.5 0.2 5.2]
 [3.4 1.4 0.2 5.2]
 [3.2 1.6 0.2 4.7]
 [3.1 1.6 0.2 4.8]
 [3.4 1.5 0.4 5.4]
 [4.1 1.5 0.1 5.2]
 [4.2 1.4 0.2 5.5]
 [3.1 1.5 0.1 4.9]
 [3.2 1.2 0.2 5. ]
 [3.5 1.3 0.2 5.5]
 [3.1 1.5 0.1 4.9]
 [3.  1.3 0.2 4.4]
 [3.4 1.5 0.2 5.1]
 [3.5 1.3 0.3 5. ]
 [2.3 1.3 0.3 4.5]
 [3.2 1.3 0.2 4.4]
 [3.5 1.6 0.6 5. ]
 [3.8 1.9 0.4 5.1]
 [3.  1.4 0.3 4.8]
 [3.8 1.6 0.2 5.1]
 [3.2 1.4 0.2 4.6]
 [3.7 1.5 0.2 5.3]
 [3.3 1.4 0.2 5. ]
 [3.2 4.7 1.4 7. ]
 [3.2 4.5 1.5 6.4]
 [3.1 4.9

In [0]:
# splitting of data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=149)

In [0]:
print("X_train",X_train)
print("X_test",X_test)

print("y_train",y_train)
print("y_test",y_test)

X_train [[2.9 4.6 1.3 6.6]
 [3.2 4.5 1.5 6.4]
 [2.4 3.7 1.  5.5]
 [3.1 5.6 2.4 6.7]
 [3.9 1.3 0.4 5.4]
 [3.8 6.7 2.2 7.7]
 [3.2 1.2 0.2 5. ]
 [2.7 5.1 1.6 6. ]
 [3.3 4.7 1.6 6.3]
 [3.  1.4 0.1 4.8]
 [2.7 4.9 1.8 6.3]
 [3.4 4.5 1.6 6. ]
 [3.4 1.4 0.3 4.6]
 [3.4 1.9 0.2 4.8]
 [3.  5.9 2.1 7.1]
 [3.  4.8 1.8 6. ]
 [3.1 4.7 1.5 6.7]
 [3.9 1.7 0.4 5.4]
 [2.8 5.1 2.4 5.8]
 [3.8 1.9 0.4 5.1]
 [2.9 3.6 1.3 5.6]
 [3.  1.3 0.2 4.4]
 [3.  1.6 0.2 5. ]
 [3.2 1.6 0.2 4.7]
 [3.6 6.1 2.5 7.2]
 [4.4 1.5 0.4 5.7]
 [3.  5.8 2.2 6.5]
 [3.2 4.8 1.8 5.9]
 [2.5 4.5 1.7 4.9]
 [2.6 4.4 1.2 5.5]
 [3.2 1.3 0.2 4.4]
 [2.5 4.  1.3 5.5]
 [3.8 1.6 0.2 5.1]
 [2.5 5.  2.  5.7]
 [3.8 1.5 0.3 5.1]
 [2.5 5.  1.9 6.3]
 [3.4 5.4 2.3 6.2]
 [3.1 1.6 0.2 4.8]
 [3.  5.5 2.1 6.8]
 [2.8 4.6 1.5 6.5]
 [3.5 1.4 0.3 5.1]
 [3.5 1.4 0.2 5.1]
 [3.4 1.5 0.4 5.4]
 [3.6 1.4 0.2 5. ]
 [3.  4.5 1.5 5.4]
 [3.4 1.6 0.2 4.8]
 [2.8 4.5 1.3 5.7]
 [4.1 1.5 0.1 5.2]
 [3.  5.1 1.8 5.9]
 [3.4 1.7 0.2 5.4]
 [3.2 1.4 0.2 4.6]
 [3.  5.8 1.6 7.2]
 [2.

In [0]:
#now building  adaboostclassifier on training sets and fitting 
classifier = AdaBoostClassifier(
    DecisionTreeClassifier(max_depth=1),
    n_estimators=100
)
classifier.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(ccp_alpha=0.0,
                                                         class_weight=None,
                                                         criterion='gini',
                                                         max_depth=1,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort='deprecated',
                          

In [0]:
#now predicting based on x_test values
predictions = classifier.predict(X_test)
predictions

array([1, 2, 2, 2, 1, 2, 2, 0, 1, 1, 0, 2, 1, 0, 0, 2, 0, 2, 0, 2, 1, 1,
       0, 2, 2, 0, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 2, 2, 2, 2, 1, 2, 2, 2,
       2])

In [0]:
y_test=y_test
predictions=predictions.reshape(-1,)


diff=pd.DataFrame({'Actual':y_test,'Predicted':predictions})
diff

Actual  Predicted
0        1          1
1        2          2
2        2          2
3        2          2
4        1          1
5        2          2
6        2          2
7        0          0
8        1          1
9        1          1
10       0          0
11       2          2
12       1          1
13       0          0
14       0          0
15       2          2
16       0          0
17       2          2
18       0          0
19       2          2
20       1          1
21       2          1
22       0          0
23       2          2
24       2          2
25       0          0
26       1          1
27       1          1
28       1          1
29       2          2
30       1          1
31       1          1
32       1          1
33       1          2
34       1          1
35       1          1
36       2          2
37       2          2
38       2          2
39       2          2
40       1          1
41       2          2
42       2          2
43       2          2
44       2          2

In [0]:
#  classification  report of our model
from sklearn.metrics import classification_report
print(classification_report(y_test,predictions))




              precision    recall  f1-score   support

           0       1.00      1.00      1.00         8
           1       0.94      0.94      0.94        16
           2       0.95      0.95      0.95        21

    accuracy                           0.96        45
   macro avg       0.96      0.96      0.96        45
weighted avg       0.96      0.96      0.96        45



In [0]:
from sklearn.metrics import mean_squared_error#calculating error of our model
mean_squared_error(diff['Actual'],diff['Predicted'])


0.044444444444444446

In [0]:
# Model Accuracy, how often is the classifier correct?
from sklearn import metrics
print("Accuracy:",metrics.accuracy_score(y_test, predictions))

Accuracy: 0.9555555555555556


# accuracy of classifier through ada boost method =**95%**

# **Decision tree model** 

In [0]:
#importing decision tree classifier
from sklearn.tree import DecisionTreeClassifier


In [0]:
#fittting the training set into model
dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)


DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [0]:

#testing the model
predict = dt.predict(X_test)
predict

array([1, 2, 2, 2, 1, 2, 2, 0, 1, 1, 0, 2, 1, 0, 0, 2, 0, 2, 0, 2, 1, 1,
       0, 2, 2, 0, 1, 1, 1, 2, 2, 2, 1, 2, 1, 1, 2, 2, 2, 2, 1, 2, 2, 2,
       2])

In [0]:
#comparing actual with prediction
y_test=y_test
predict=predict.reshape(-1,)

diff=pd.DataFrame({'Actual':y_test,'Predicted':predict})
diff

Actual  Predicted
0        1          1
1        2          2
2        2          2
3        2          2
4        1          1
5        2          2
6        2          2
7        0          0
8        1          1
9        1          1
10       0          0
11       2          2
12       1          1
13       0          0
14       0          0
15       2          2
16       0          0
17       2          2
18       0          0
19       2          2
20       1          1
21       2          1
22       0          0
23       2          2
24       2          2
25       0          0
26       1          1
27       1          1
28       1          1
29       2          2
30       1          2
31       1          2
32       1          1
33       1          2
34       1          1
35       1          1
36       2          2
37       2          2
38       2          2
39       2          2
40       1          1
41       2          2
42       2          2
43       2          2
44       2          2

In [0]:
#  classification  report of our model
from sklearn.metrics import classification_report
print(classification_report(y_test,predict))



              precision    recall  f1-score   support

           0       1.00      1.00      1.00         8
           1       0.93      0.81      0.87        16
           2       0.87      0.95      0.91        21

    accuracy                           0.91        45
   macro avg       0.93      0.92      0.93        45
weighted avg       0.91      0.91      0.91        45



In [0]:
#calculating error
from sklearn.metrics import mean_squared_error
mean_squared_error(diff['Actual'],diff['Predicted'])


0.08888888888888889

In [0]:
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, predict))

Accuracy: 0.9111111111111111


# accuracy of classifier through Decision tree method =**91%**

# **Report**

**from the above we find the accuracy of the model is good by using ensemble method (ada boost)  then generalclassifier like  svm , decissiontree ,kNNclassifier etc**




